crates/ruff_db/src/parsed.rs

```diff
@@ -0,0 +1,99 @@
+use std::fmt::Formatter;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use ruff_python_ast::{ModModule, PySourceType};
+use ruff_python_parser::{parse_unchecked_source, Parsed};
+
+use crate::source::source_text;
+use crate::vfs::{VfsFile, VfsPath};
+use crate::Db;
+
+/// Returns the parsed AST of `file`, including its token stream.
+///
+/// The query uses Ruff's error resilient parser. That means that the parser always succeeds to produce a
+/// AST even if the file contains syntax errors. The syntax errors are Parsing the module succeeds even when the file contains syntax error. The parse errors
+/// are then accessible through [`Parsed::errors`].
+///
+/// The parse tree is cached between invocations, but the query doesn't make use of Salsa's optimization
+/// that skips dependent queries if the AST hasn't changed. Comparing two ASTs is a non-trivial operation
+/// and every offset change is directly reflected in the changed AST offsets. Ruff's AST also doesn't implement `Eq`.
+/// which is required to use the optimization.
+#[salsa::tracked(return_ref, no_eq)]
+pub fn parsed_module(db: &dyn Db, file: VfsFile) -> Parsed<ModModule> {
+    let source = source_text(db, file);
+    let path = file.path(db);
+
+    let ty = match path {
+        VfsPath::FileSystem(path) => path
+            .extension()
+            .map_or(PySourceType::Python, PySourceType::from_extension),
+        VfsPath::Vendored(_) => PySourceType::Stub,
+    };
+
+    parse_unchecked_source(&source, ty)
+}
+
+/// Cheap cloneable wrapper around the parsed module.
+#[derive(Clone, PartialEq)]
+pub struct ParsedModule {
+    inner: Arc<Parsed<ModModule>>,
+}
+
+impl ParsedModule {
+    /// Consumes `self` and returns the Arc storing the parsed module.
+    pub fn into_arc(self) -> Arc<Parsed<ModModule>> {
+        self.inner
+    }
+}
+
+impl Deref for ParsedModule {
+    type Target = Parsed<ModModule>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl std::fmt::Debug for ParsedModule {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("ParsedModule").field(&self.inner).finish()
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Tests also for stub file, vendored stub file, and unknown-extension file?

Reply:
I added a test for vendored paths. I prefer to not add a test for unknown extensions and stub files because I then end up re-testing `PySourceType::from_extension`. I want to add integration tests for these files.